hw/xbox/xid-sdl.c

```diff
@@ -195,35 +193,14 @@ static const XIDDesc desc_xid_xbox_gamepad = {
 #define BUTTON_MASK(button) (1 << ((button) - GAMEPAD_DPAD_UP))
 
 static void update_output(USBXIDState *s) {
-#ifdef FORCE_FEEDBACK
     if (s->sdl_haptic == NULL) { return; }
+    if (s->sdl_haptic_effect == NULL) { return; }
 
-    SDL_HapticLeftRight effect = {
-        .type = SDL_HAPTIC_LEFTRIGHT,
-        .length = SDL_HAPTIC_INFINITY,
-        /* FIXME: Might be left/right inverted */
-        .large_magnitude = s->out_state.right_actuator_strength,
-        .small_magnitude = s->out_state.left_actuator_strength
-    };
-
-    if (s->sdl_haptic_effect_id == -1) {
-        int effect_id = SDL_HapticNewEffect(s->sdl_haptic,
-                                            (SDL_HapticEffect*)&effect);
-        if (effect_id == -1) {
-            fprintf(stderr, "Failed to upload haptic effect!\n");
-            SDL_HapticClose(s->sdl_haptic);
-            s->sdl_haptic = NULL;
-            return;
-        }
-        SDL_HapticRunEffect(s->sdl_haptic, effect_id, 1);
-
-        s->sdl_haptic_effect_id = effect_id;
+    s->sdl_haptic_effect->type = SDL_HAPTIC_LEFTRIGHT;
```

Comment:
I have to re-set the type each time we update the controller. Something within SDL is changing this value to `NULL`. Haven't narrowed it down yet, but SDL manipulating this memory feels wrong, so maybe this isn't the recommended way of using the API?

I will keep digging on this.

Reply:
See https://github.com/xqemu/xqemu/pull/197/files#r255324191

I still think you can just have a new local variable here. If you feel like duplicating the init of 2 fields is too much, you could just move it into a setter for the SDL structure like `SetLeftRightHapticEffect(&effect, left_strength, right_strength)` (maybe SDL already has macros / functions for this?)